### PR TITLE
Fix an error in the oneDPL `__future` implementation

### PR DESCRIPTION
**Pull request by SergeyKopienko**
In this PR we fix an error in the oneDPL `__future` implementation.

Let's take look to current implementation of `__future::wait()` function :
```C++
    wait()
    {
#if !ONEDPL_ALLOW_DEFERRED_WAITING
        __my_event.wait_and_throw();
#endif
    }
```
This mean that real call of `wait_and_throw()` will depends on the state of `ONEDPL_ALLOW_DEFERRED_WAITING` macro (what is this and for that - described at https://oneapi-src.github.io/oneDPL/macros.html).

Now let's take a look to implementation of `__future::__wait_and_get_value()` function :
```C++
    template <typename _T>
    constexpr auto
    __wait_and_get_value(_T& __val)
    {
        wait();
        return __val;
    }
```

So at first - we waits and then return value. But this mean that real call of `wait_and_throw()` will depends on the state of `ONEDPL_ALLOW_DEFERRED_WAITING` macro too and we may return value without real `wait_and_throw()` call.

### CODE REVIEW

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:701]: Do we want users to be encouraged to call a `deferrable_wait()` function?  I would think that is somewhat implementation details or perhaps to convoluted for public consumption.
  
Obviously we need this to be `public`, but should we uglify it maybe to discourage usage outside the library (or would we want some internal version of `__future` even which we don't return to the users?

File: test/general/header_inclusion_order_async_0.pass.cpp
Comment (danhoeflinger) [inline test/general/header_inclusion_order_async_0.pass.cpp:37]: As mentioned above, I'm not sure we want users calling this function, but rather they would be calling `wait()`

File: test/general/header_inclusion_order_async_1.pass.cpp
Comment (danhoeflinger) [inline test/general/header_inclusion_order_async_1.pass.cpp:37]: Same as above, users probably should call `wait()`

**Review by danhoeflinger — COMMENTED** [review]
I'm wondering if this macro is still used by anyone, and if we still need / want to support it.  It seems like the more appropriate venue for something like this is to use the async API.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:701]: Our `__future` implemented in the namespace 
```C++
namespace oneapi
{
namespace dpl
{
namespace __internal
{
```
This mean (for me) that user will never works with our `__future` implementation directly.

File: test/general/header_inclusion_order_async_0.pass.cpp
Comment (SergeyKopienko) [reply test/general/header_inclusion_order_async_0.pass.cpp:37]: @akukanov how do you think, which call we should use in this case?

File: test/general/header_inclusion_order_async_1.pass.cpp
Comment (SergeyKopienko) [reply test/general/header_inclusion_order_async_1.pass.cpp:37]: @akukanov how do you think, which call we should use in this case?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h
Comment (akukanov) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:701]: Actually they might, if an object of this type is returned by our public API functions. It might be documented as an unspecified type, yet it should be considered as a part of public API. Is that the case with `__future`?

File: test/general/header_inclusion_order_async_0.pass.cpp
Comment (SergeyKopienko) [reply test/general/header_inclusion_order_async_0.pass.cpp]: Fixed: restored `wait()` call.

File: test/general/header_inclusion_order_async_1.pass.cpp
Comment (SergeyKopienko) [reply test/general/header_inclusion_order_async_1.pass.cpp]: Fixed: restored `wait()` call.

File: test/general/header_inclusion_order_async_0.pass.cpp
Comment (danhoeflinger) [reply test/general/header_inclusion_order_async_0.pass.cpp]: I recommend at least uglifying it, but we could consider having an `__internal_future` type with this functionality, and another which we return to users which can be created from an `__internal_future` which does not have this functionality.

**Comment by julianmi** [comment]
> I'm wondering if this macro is still used by anyone, and if we still need / want to support it. It seems like the more appropriate venue for something like this is to use the async API.

I agree with this.

If we want to keep it, I think we should improve the documentation for `ONEDPL_ALLOW_DEFERRED_WAITING` in https://github.com/oneapi-src/oneDPL/blob/main/documentation/library_guide/macros.rst. It is currently very unspecific especially regarding the need for synchronization and potential data issues:
"This macro allows waiting for completion of certain algorithms executed with device policies to be deferred. (Disabled by default.)"

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h]: Yes, its an unspecified future-like type in our documentation, which lists `wait()` and `get()`.  It of course is not in the spec because it is part of the experimental async API.

File: include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h
Comment (akukanov) [inline include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h:744]: I am concerned that deferring the wait might not be appropriate here, in particular if the input for "Round 2" depends on completion of the kernel(s) above it.

File: test/general/header_inclusion_order_async_0.pass.cpp
Comment (SergeyKopienko) [reply test/general/header_inclusion_order_async_0.pass.cpp]: Uglified.

File: include/oneapi/dpl/pstl/hetero/numeric_impl_hetero.h
Comment (akukanov) [inline include/oneapi/dpl/pstl/hetero/numeric_impl_hetero.h:178]: Here I have the same kind of concern as well, as another computation follows the pattern call.

**Comment by akukanov** [comment]
> I'm wondering if this macro is still used by anyone, and if we still need / want to support it. It seems like the more appropriate venue for something like this is to use the async API.

@danhoeflinger, you are right of course. However, the async API is still in preview, and the macro is public so formally needs deprecation.
- 
> If we want to keep it, I think we should improve the documentation for ONEDPL_ALLOW_DEFERRED_WAITING in https://github.com/oneapi-src/oneDPL/blob/main/documentation/library_guide/macros.rst. It is currently very unspecific especially regarding the need for synchronization and potential data issues.

Yes, something might be added along the lines that callers are responsible for waiting for completion on the SYCL queue associated with the execution policy before accessing any data modified by deferred algorithms.

**Comment by danhoeflinger** [comment]
> @danhoeflinger, you are right of course. However, the async API is still experimental, and the macro is public so formally needs deprecation.


Yes, you are right, it would require deprecation.  Should we consider adding deprecation for the next release?  In my opinion the experimental async API is already a much better supported and better documented version of what this macro provides.

File: include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h]: Fixed

File: include/oneapi/dpl/pstl/hetero/numeric_impl_hetero.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/numeric_impl_hetero.h]: Fixed

**Comment by akukanov** [comment]
I think we should first have a plan for async API productization, then talk about deprecation.

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:57]: I cant currently find any problems with this one currently in the repo, but it is to some extent a "building block", which other algorithms may use internally as a piece of a multi-kernel operation which may have internal dependencies.  

This seems like a possible source of errors future the way it is currently implemented.
It may be worth it to do something like what is done in `__patern_walk2` to provide control of `_IsSync` to the caller, then use `ONEDPL_ALLOW_DEFERRED_WAITING` to control that template variable in the cases which make sense.  Alternatively, we could allow algorithms to explicitly set this pattern to synchronous, and otherwise, by default depend on `__deferrable_wait()`.

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:104]: `__pattern_partial_sort_copy` uses `__pattern_walk2` in a way which requires synchronousity.  It needs the copy to finish before sort is started on the copied data.    
Also, this function has a template parameter which specifies `_IsSync`, and then we ignore it by using `__deferrable_wait()`.  Should we instead selectively use the macro to to control the `_IsSync` template parameter (only when it is safe to do so)?

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:161]: similar comment to `__pattern_walk1`.

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:1188]: `__pattern_inplace_merge` requires this pattern to by synchronous, so the merge is finished before the copy.

**Review by danhoeflinger — COMMENTED** [review]
After investigating a handful of these individual patterns, it makes me question if this is the right strategy.

It should be said that we are not adding errors where things were working before.  This is already broken when the macro is enabled.  However, if we are looking at improving this, we should try to fix these sorts of errors and "traps" for future bugs.  

What we actually want from this macro is optional synchronicity at the API level, but we do not have a consistent standard or expectation for synchronicity at the pattern level, which can be problematic.

@reble This PR is probably something to consider in our look at aligning with the asynchronous API.  Even if it is effectively a more crude precursor to that API.

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h
Comment (akukanov) [reply include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:104]: @SergeyKopienko please pay attention to this.

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h
Comment (akukanov) [reply include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:1188]: @SergeyKopienko please take into account.

**Comment by danhoeflinger** [comment]
> @danhoeflinger @julianmi While this patch might still leave some problems unresolved, I think at this time it's an improvement - first, due to limiting the number of places where waiting can be deferred and making these visible and searchable in the code, and second, due to improving the documentation for the macro. Therefore I would still propose to commit it, maybe with adding some TODOs and/or an issue to revisit later.

OK, do we want to try to resolve issues with the individual usages like https://github.com/oneapi-src/oneDPL/pull/1747#discussion_r1710009584 in this PR, or leave those as TODO?  
If we want to try to resolve them, I can keep looking in detail through the rest of the patterns.

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:104]: @danhoeflinger, @akukanov Just for note - this error hasn't been introduced in this PR.

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:211]: @danhoeflinger please explain which chain of call exactly you mean.

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:1238]: @akukanov, @danhoeflinger Fixed.

File: documentation/library_guide/macros.rst
Comment (julianmi) [inline documentation/library_guide/macros.rst:105]: There seems to be a line break in the rendered version between "waiting" and "for".

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:700]: Do we need to wait here again even if the user has called `__future.wait()`? Could we instead check for `__waited_for_result` and skip the wait if it is true.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:691]: Similar comment here about unneeded waits in case we've waited before.

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:154]:
```suggestion
    // Call no wait, wait or deferrable wait depending on _WaitMode
```

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:700]: My assumption that `__waited_for_result` state is debugging tool only, may be we remove it at all later or place under some `debug` mode macro check. So I believe that we shouldn't use it for choice how to get result and so on. With this tool I am trying to find errors in tests now on `CPU` and on `GPU`.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:691]: https://github.com/oneapi-src/oneDPL/pull/1747#discussion_r1715079486

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h]: Fixed.

File: documentation/library_guide/macros.rst
Comment (SergeyKopienko) [reply documentation/library_guide/macros.rst:105]: Fixed.

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:210]: UPD: all tests passed on `CPU` and `GPU` devices and no `assert` has been occurred,
I mean `assert` from
```C++
    template <typename _T>
    constexpr auto
    __wait_and_get_value(const _T& __val)
    {
        wait();

        assert(__get_waited_for_result() && "Result was not waited");

        return __val;
    }
```
So we may have hope that the current state of the code is correct.

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:1562]: I believe this can stay `__async_mode` based on the comment below.  
The temporary buffer will order the kernels appropriately, and no explicit wait is required. (Also, since `__buf_last` does not depend on the result of the kernel, it is a fixed increment of `__buf_first`)

File: include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h:519]: I originally thought  `__pattern_inplace_merge` could have an issue, but I believe it is saved by the implicit dependency between kernels from the temporary buffer, so I think nothing needs to be done.

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h]: Fixed

File: include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h:519]: ok

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h]: as result, all `waited_for_result` stuff has been removed to reduce the amount of changes.

Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h]: @julianmi may we resolve this conversation?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h]: as result, all `waited_for_result` stuff has been removed to reduce the amount of changes.

Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h]: @julianmi may we resolve this conversation?

File: include/oneapi/dpl/internal/async_impl/async_impl_hetero.h
Comment (danhoeflinger) [inline include/oneapi/dpl/internal/async_impl/async_impl_hetero.h:74]: I believe this is the wrong variable name `__future` vs `__future_obj`

Comment (SergeyKopienko) [reply include/oneapi/dpl/internal/async_impl/async_impl_hetero.h]: Fixed

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h
Comment (akukanov) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:772]: Do we really need to make this dispatch generic, for an "arbitrary" `_Future` type? Or would it make more sense to have that dispatch being really within the actual future type used in the implementation?

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:162]: We discussed and think its best to leave this for now prior to release, and perhaps make more changes after release to switch some patterns to a system like `__pattern_walk2`.

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:57]: We discussed and think its best to leave this for now prior to release, and perhaps make more changes after release to switch some patterns to a system like __pattern_walk2.

**Review by danhoeflinger — APPROVED** [review]
Wait for a Green CI, and please resolve the feedback from @akukanov , but overall LGTM.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h
Comment (akukanov) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:772]: Something like this, for example (not tested):
```
// within class template __future
template<typename _ModeTag>
void wait(_ModeTag)
{
    if constexpr (std::is_same_v<_ModeTag,__sync_mode>)
        wait();
    else if constexpr (std::is_same_v<_ModeTag,__deferrable_mode>)
        __deferrable_wait();
}
```

Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h]: Fixed

File: include/oneapi/dpl/internal/async_impl/async_impl_hetero.h
Comment (akukanov) [inline include/oneapi/dpl/internal/async_impl/async_impl_hetero.h:51]: Do we really need this tag for `__pattern_walk2_async`? I mean, it is supposed to be asynchronous, is not it? And of course currently it is not used with other than the default template argument.

Comment (SergeyKopienko) [reply include/oneapi/dpl/internal/async_impl/async_impl_hetero.h:51]: The source idea was to have the same format for `__pattern_walk2_async` and for `__pattern_walk2`

Comment (akukanov) [reply include/oneapi/dpl/internal/async_impl/async_impl_hetero.h:51]: I know; and I know it is not in your patch. But my question is, is it really needed? And we are changing the pattern slightly anyway, so I think it's appropriate to revisit this.

Comment (SergeyKopienko) [reply include/oneapi/dpl/internal/async_impl/async_impl_hetero.h:51]: From current use cases state it's really not required.

Comment (SergeyKopienko) [reply include/oneapi/dpl/internal/async_impl/async_impl_hetero.h]: Fixed: removed.

**Review by akukanov — APPROVED** [review]

**Review by danhoeflinger — APPROVED** [review]
It would be best to wait for green CI, LGTM.

**Review by julianmi — DISMISSED** [review]

**Review by julianmi — APPROVED** [review]

**Pull request merged**
